kernel/univ.ml

```diff
@@ -515,71 +376,63 @@ struct
       let v' = f v in 
 	if v' == v then x
 	else if Level.is_prop v' && n != 0 then
-	  hcons (Level.set, n)
-	else hcons (v', n)
+	  (Level.set, n)
+	else (v', n)
 
   end
-    
-  let compare_expr = Expr.compare
 
-  module Huniv = HList.Make(Expr.HExpr)
-  type t = Huniv.t
-  open Huniv
-    
-  let equal x y = x == y || 
-    (Huniv.hash x == Huniv.hash y && 
-       Huniv.for_all2 Expr.equal x y)
+  type t = Expr.t list
 
-  let hash = Huniv.hash
+  let tip l = [l]
+  let cons x l = x :: l
 
-  let compare x y =
-    if x == y then 0
-    else 
-      let hx = Huniv.hash x and hy = Huniv.hash y in
-      let c = Int.compare hx hy in 
-	if c == 0 then
-	  Huniv.compare (fun e1 e2 -> compare_expr e1 e2) x y
-	else c
+  let rec hash = function
+  | [] -> 0
+  | e :: l -> Hashset.Combine.combinesmall (Expr.ExprHash.hash e) (hash l)
+
+  let equal x y = x == y || List.equal Expr.equal x y
+
+  let compare x y = if x == y then 0 else List.compare Expr.compare x y
+
+  module Huniv = Hashcons.Hlist(Expr)
 
-  let rec hcons = function
```

Comment:
There's something I don't understand here, why did we have a re-hashconsing function? That was used at Qed AFAIU.

Reply:
Because hashconsing in Coq is not hashconsing! In practice, because of marshalling, you lose the property that two equivalent objects are physically equal. Thus it is always safer to rehashcons things when possible.

Reply:
Ok, that was my guess as well. However it means we're really doing hashconsing twice or more on already hashconsed objects...

Reply:
And is there any reason we do that only on universes? As opposed to names and such?

Reply:
What? No, we rehashcons everybody (unluckily). That's why it's damn slow!

Reply:
I don't understand. Do we rehashcons everything after unmarshalling? Also, can you confirm we never rely on maximal sharing for correctness?

Reply:
No.

1. We hashcons everything *before* marshalling. After marshalling, screw us, the unicity property is lost. In practice, we marshal before storing terms in the global environment when we pass through the kernel.
2. We never ever EVAR rely on maximal sharing for correctness.
3. EVAR does not stand for evar.